Allow `icalcomponent_foreach_recurrence` to receive `DATE`-only `start` and `end` params.

minichma opened the pull request:
Fixes #833.

* Allow `icalcomponent_foreach_recurrence` to receive `DATE`-only `start` and `end` params.
* Fix incorrect calculation of an event's duration in case `DTSTART`/`DTEND` are DATE-only.
* Mention limitation that the lib doesn't distinguish between nominal and exact event durations (related to #630).

winterz reviewed (approved):
@psychon please test

once this lands I will cherrypick to the 3.0 branch for the upcoming 3.0.19 release

minichma commented:
@winterz As discussed in #633 I modified `icaltime_span_new()` in that I removed the special handling of the case that `end` is unset. Within libical this case doesn't occur, which probably is the reason for the problem not popping up more frequently. However, in v3 this change might be breaking for certain users. On the other hand, as it was broken anyways, I wouldn't expect real life problems. What do you think? Alternatively we could just leave this aspect as it was, as it doesn't do any harm either.

winterz commented:
> @winterz As discussed in #633 I modified `icaltime_span_new()` in that I removed the special handling of the case that `end` is unset. Within libical this case doesn't occur, which probably is the reason for the problem not popping up more frequently. However, in v3 this change might be breaking for certain users. On the other hand, as it was broken anyways, I wouldn't expect real life problems. What do you think? Alternatively we could just leave this aspect as it was, as it doesn't do any harm either.
> 
> [Edit] ... or ignore the last commit when cherry-picking into v3.

we need to deal with any pain now (in the new v3 with this added) or later in 4.0
may as well go for it now with v3

winterz commented:
I'm happy if you do the merge
can you also do the cherry-pick into 3.0?

minichma commented:
Certainly. Done.
